### PR TITLE
Use spf13/cobra for `docker run` and `docker create`

### PR DESCRIPTION
**Pull request by dnephin**
This PR builds off of #22769

As requested in https://github.com/docker/docker/pull/22769#issuecomment-222833350 this is a PoC to show that cobra and pflag will work for a more involved command like `docker run`.

### CODE REVIEW

**Comment by thaJeztah** [comment]
ping @duglin @LK4D4 @vdemeester PTAL

**Comment by vdemeester** [comment]
Code looks good :+1: (and I like the `TODO`s on the flags :angel:)
Tried it a little bit, seems fine but didn't have time for a complete test of the `run` command.

**Comment by thaJeztah** [comment]
Hm, looks like I found a really weird issue;

This fails:
```
docker run -v /bla/bla:/foo/bar:rw,Z -dit -p 127.0.0.1:80:80 alpine sh
```

So does this:
```
docker run -it --rm alpine sh
```

But using "busybox" works;

```
docker run -v /bla/bla:/foo/bar:rw,Z -dit -p 127.0.0.1:80:80 busybox sh
```

```
docker run -it --rm busybox sh
```

For alpine it sends an empty command;

```
{
  "AttachStderr": false,
  "AttachStdin": false,
  "AttachStdout": false,
  "Cmd": null,
  "Domainname": "",
  "Entrypoint": null,
  "Env": [
    
  ],
..
```

**Comment by vdemeester** [comment]
Hum some weirdies though (tried it out with @thaJeztah ) : 

```
root@fb4c2eda9e89:/go/src/github.com/docker/docker# ./bundles/latest/binary-client/docker run -v /var:/foo:rw,Z -dit -p 127.0.0.1:80:80 busybox sh
b5424362a55dc975347e3ebb097db45ba8389f69d84e00586e230b50160af160
root@fb4c2eda9e89:/go/src/github.com/docker/docker# ./bundles/latest/binary-client/docker run -v /var:/foo:rw,Z -dit -p 127.0.0.1:80:80 alpine sh
Unable to find image 'alpine:latest' locally
latest: Pulling from library/alpine
d0ca440e8637: Pull complete 
Digest: sha256:9dd95238d7f6ac0fabafd7b52e52d8a7c63a5e5b390d27e13cb5da8ddd2f0d98
Status: Downloaded newer image for alpine:latest
./bundles/latest/binary-client/docker: Error response from daemon: No command specified.
See './bundles/latest/binary-client/docker run --help'.
root@fb4c2eda9e89:/go/src/github.com/docker/docker# ./bundles/latest/binary-client/docker run -v /var:/foo:rw,Z -dit -p 127.0.0.1:80:80 alpine sh
./bundles/latest/binary-client/docker: Error response from daemon: No command specified.
See './bundles/latest/binary-client/docker run --help'.
```

Looks like the command is not passed (`busybox` has a default command, `alpine` does not)

**Comment by thaJeztah** [comment]
Ok we (@vdemeester) found the difference; it appears to fail if there's no default command in the image. Doing this works:

```
FROM alpine
CMD sh
```

```
docker build -t frommie .
docker run -it --rm frommie sh
```

**Comment by thaJeztah** [comment]
Some more testing, and this works ok

```
export FOOBAR=blabla

docker run \
    -v /bla/bla:/foo/bar:rw,Z \
    -dit \
    -p 127.0.0.1:80:80 \
    -v $(docker volume create --name foobar):/foobar \
    -u 123:456 \
    --label bla=foo \
    -l bla=bar \
    -e FOOBAR \
    -e something=something \
    --memory 300m \
    --security-opt apparmor:unconfined \
    --security-opt seccomp=unconfined \
    --name cobra \
    busybox sh
```

**Comment by dnephin** [comment]
Oops, sorry about that. I had a fix for that issue and I forgot to push it before opening the PR.

That issue is now fixed.

**Comment by dnephin** [comment]
`docker create` is implemented now as well, so more of the tests should pass

File: api/client/container/run.go [hunk 1]
```diff
@@ -270,32 +268,58 @@ func (cli *DockerCli) CmdRun(args ...string) error {
...
 	if status != 0 {
-		return Cli.StatusError{StatusCode: status}
+		return cli.StatusError{StatusCode: status}
 	}
 	return nil
 }
+
+// reportError is a utility method that prints a user-friendly message
+// containing the error that occurred during parsing and a suggestion to get help
+func reportError(stderr io.Writer, name string, str string, withHelp bool) {
```
Comment (duglin) [inline api/client/container/run.go]: Is this something that should be in a generic spot so it can be reused by other commands?

Comment (dnephin) [inline api/client/container/run.go]: Doesn't seem to be used anywhere else. I just moved this, it already existed before my change.

Comment (duglin) [inline api/client/container/run.go]: hmm interesting.

Comment (dnephin) [inline api/client/container/run.go]: I think this is trying to mimic the error message that is normally raised when a flag or value is incorrect. These two commands do a bunch of validation on the flag values after the initial parsing, where as I think many commands don't do any additional validation.

**Comment by dnephin** [comment]
The build is green!

**Comment by cpuguy83** [comment]
needs rebase

**Comment by thaJeztah** [comment]
ow; flaky/faulty test on WindowsTP5 (not related); guess this is the issue with Windows not liking filenames named `#1`

```
17:54:09 ----------------------------------------------------------------------
17:54:09 FAIL: docker_cli_build_test.go:6760: DockerSuite.TestBuildDockerignoreComment
17:54:09 
17:54:09 docker_cli_build_test.go:6791:
17:54:09     c.Fatal(err)
17:54:09 ... Error: failed to build the image: Sending build context to Docker daemon 5.632 kB

17:54:09 Step 1 : FROM busybox
17:54:09  ---> 3f1b2939d14c
17:54:09 Step 2 : ADD . /tmp/
17:54:09  ---> 78175a922461
17:54:09 Removing intermediate container e893b2ece047
17:54:09 Step 3 : RUN sh -c "(ls -la /tmp/#1)"
17:54:09  ---> Running in 14e3ac5d6f7e
17:54:09 The command 'cmd /S /C sh -c "(ls -la /tmp/#1)"' returned a non-zero code: 255
```

**Comment by thaJeztah** [comment]
Opened https://github.com/docker/docker/issues/23246 for the test

**Comment by thaJeztah** [comment]
Windows test will be fixed in https://github.com/docker/docker/issues/23248

**Comment by dnephin** [comment]
it's green!

**Comment by thaJeztah** [comment]
LGTM

**Comment by vdemeester** [comment]
Arf, needs a rebase :sweat:

**Comment by thaJeztah** [comment]
Ah, probably because this one was merged https://github.com/docker/docker/pull/23241. I hoped this one would get merged first

**Comment by vdemeester** [comment]
LGTM :angel: 
Let's carry this to merge it quick :stuck_out_tongue_closed_eyes:

**Pull request closed without merging**